File: sdk/python/src/anyvali/schemas/never.py
"""Never schema."""

from __future__ import annotations

from typing import Any

from ..issue_codes import INVALID_TYPE
from .base import BaseSchema, ValidationContext


class NeverSchema(BaseSchema):
    """Schema that rejects all values."""

    def __init__(self) -> None:
        super().__init__()

    def _validate(self, input: Any, ctx: ValidationContext) -> Any:
        ctx.add_issue(INVALID_TYPE, "No value is allowed", expected="never", received=type(input).__name__)
        return None

    def _to_node(self) -> dict[str, Any]:
        return self._add_common_node_fields({"kind": "never"})

File: sdk/js/src/schemas/never.ts
import type { ParseContext, SchemaNode } from "../types.js";
import { BaseSchema } from "./base.js";
import { ISSUE_CODES } from "../issue-codes.js";
import { describeType } from "../util.js";

export class NeverSchema extends BaseSchema<never, never> {
  _validate(input: unknown, ctx: ParseContext): unknown {
    ctx.issues.push({
      code: ISSUE_CODES.INVALID_TYPE,
      message: `Expected never (no value is valid)`,
      path: [...ctx.path],
      expected: "never",
      received: describeType(input),
    });
    return undefined;
  }

  _toNode(): SchemaNode {
    return { kind: "never" } as SchemaNode;
  }
}
